### PR TITLE
'Realign' IR trees via the intermediary buffer

### PR DESCRIPTION
**Pull request by tybug**
This brings back the status quo of buffer shrinking, where misalignments carry on as a new interpretation of their new type. E.g. `integers() | text()` shrinks from `[1, s: str]` to `[0, n: int]`, and the bytes for s got interpreted as some integer n, instead of giving up (what we do in the ir based shrinker in master). The reason this works well - better than randomly generating n  - is the buffer encodes an ordering, so you guarantee an 'equivalently complex' integer n, and avoid increasing complexity via misalignments.

We're actively trying to move *off* the buffer, though! So this pull is a temporary stepping stone and we'll eventually need to create a native ir ordering. But the realigning principle will be similar.

This resolves some shrinking regressions (#4006), especially in stateful testing, where misalignment is common (consider decreasing the integer choosing the rule to execute, where rules may use various strategies). And it improves shrink calls (via `hypothesis-python/benchmark`):

image

This also removes the notion of a data being "invalid" after a misalignment. I think this is too complicated and asking for trouble. I'd like to always generate *something* for misalignments, like we used to. (I'm a bit sad to throw away the effort I put into tests + etc here - but not too much, since I am to this day not confident the eg caching logic is correct for invalid datas).

### CODE REVIEW

File: hypothesis-python/tests/common/debug.py [hunk 1]
```diff
@@ -18,7 +18,7 @@
 
 from tests.common.utils import no_shrink
 
-TIME_INCREMENT = 0.01
+TIME_INCREMENT = 0.001
```
Comment (tybug) [inline hypothesis-python/tests/common/debug.py:21]: `_consistently_increment_time` was giving me trouble with spurious deadline errors, so I've just bumped our increment down so we don't hit 200ms. The misalignment logic hits more time.time() calls than before via constructing ConjectureDatas (`self.start_time = time.perf_counter()`).

**Comment by tybug** [comment]
well, those are some interesting failures. I wonder if `generate_mutations_from` is turning up new interactions as it inherits this realigning behavior 🤔

**Comment by Zac-HD** [comment]
> well, that's [an interesting django failure](https://github.com/HypothesisWorks/hypothesis/actions/runs/10399472727/job/28798400738?pr=4086). I wonder if `generate_mutations_from` is turning up new interactions as it inherits this realigning behavior 🤔

Proximate cause looks pretty standard to me - almost certainly a datetime near the beginning of year zero, such that converting it into the timezone would put it earlier than Python can represent.

Probably worth reporting this to Django in case they want to do something upstream, but then we could avoid it by adding a 24h buffer at each end [here](https://github.com/HypothesisWorks/hypothesis/blob/b38fe7a6e475008ff4341de2ae984aac8e5e9f43/hypothesis-python/src/hypothesis/extra/django/_fields.py#L114-L119).

**Comment by tybug** [comment]
You caught me about 3 minutes before I hit submit on a new issue for this 😄 That is indeed the cause. They convert to naive due to e.g. storing in sqlite which has no timezone aware support, so if we have a close-to-datetime.min aware datetime, it can underflow. I'll report to django, but it's unclear to me they will or can do anything here.

**Comment by tybug** [comment]
Filed over at django: https://code.djangoproject.com/ticket/35683. I'll constrain the range within 24 hours for now like you suggested.

**Review by Zac-HD — APPROVED** [review]

**Pull request merged**